### ISSUE TITLE
Context parameters instead of Monad

### ISSUE THREAD

**Issue by pjljvandelaar** [TorXakis/TorXakis#361]
We use Monads to pass variables around that are only read at some points.
We are experiencing quite some issues with combining Monads.
I propose to introduce a context parameter.
We need to pass a single parameter around, and
at the needed locations get the info from this data structure.
I am still in doubt whether the (run-time) settings are part of this context parameter or it is a separate parameter.
Maybe it just depends on the situation.....

**Comment by capitanbatata** [comment]
Can you give an example of the issues we're experiencing with combining monads?

**Comment by pjljvandelaar** [comment]
@tretmans could you describe you issues with call back functions and the different monads?

**Comment by pjljvandelaar** [comment]
For `subst` I propose the following solution:

Pass around an `environment` which inherits from the some of the following classes
```
HasSorts
HasFuncs
HasProcs
```
For example,
```
class HasFuncs 
   where funcdefs
```

**Comment by pjljvandelaar** [comment]
@capitanbatata what do you think of only three environments:
`SortEnv` only `HasSorts`
`FuncEnv` has `HasSorts` and `HasFuncs`
`ProcEnv` has all?

Advantage: data structure remains flat and overloading is allowed.

Alternative (with more nesting of data):
`SortEnv`  `HasSorts`
`FuncEnv` contains a `SortEnv` and `HasFuncs`
`ProcEnv` contains a `FuncEnv` and `HasProcs`

**Comment by capitanbatata** [comment]
That could work as well. The disadvantage is that we would need to implement the instances multiple times.

For instance:

```haskell
instance HasSorts SortEnv where 
   -- blah

instance HasFuncs FuncEnv where
   -- blah blah

instance HasSorts FuncEnv where
   -- more blah
```

And we would need three more instances (with a lot of boilerplate as well) for `ProcEnv`. Also we're imposing a more rigid structure it seems.

All in all, this approach and mine have their advantages and disadvantages, so it is a matter of doing some mocking up, playing around, and seeing what's the most convenient for `TorXakis`. You could experiment with your idea by implementing something (or re-implementing the `txs-compiler` in an experimental branch), and so we can compare the different solutions.

A third alternative would be using some type-level computation using heterogeneous lists. I'll try to give this a try when I have some free time.

**Comment by pjljvandelaar** [comment]
Yeah, it is unfortunate that we can't provide a default implementation...
Since for all three indeed we want the same implementation....

**Comment by capitanbatata** [comment]
Ok, after asking around in SO, I got [this excellent answer](https://stackoverflow.com/questions/49545045/defining-incremental-environments-without-running-into-overlapping-instances/). We can definitely use it for defining what I call incremental environments. Furthermore, since these ideas can be also used at the compiler, we can promote the `Has` class to a library that we can use in `TorXakis`.

**Comment by pjljvandelaar** [comment]
Nice answer... higher Haskell....

**Comment by capitanbatata** [comment]
Well actually type level programming in Haskell. But it is not that difficult actually, and the type computations that we need are quite trivial.

**Comment by pjljvandelaar** [comment]
Let's sit together and build it for hasSort, hasFunction, and hasProcs

**Comment by capitanbatata** [comment]
Sure.
